app/models/concerns/unparsed_run.rb

```diff
@@ -13,154 +18,286 @@ def parse_into_db
+
+      (0...run.attempt_history_len).each do |i|
+        attempt = run.attempt_history_index(i)
+        dates = [attempt.started, attempt.ended]
+
+        history = RunHistory.new(
+          run_id:               id,
+          attempt_number:       attempt.index.to_i,
+          realtime_duration_ms: attempt.time.real_time&.total_seconds&.*(1000) || nil,
+          gametime_duration_ms: attempt.time.game_time&.total_seconds&.*(1000) || nil,
+          pause_duration_ms:    attempt.pause_time&.total_seconds&.*(1000) || 0,
+          started_at:           dates[0] && DateTime.parse(dates[0].to_rfc3339),
+          ended_at:             dates[1] && DateTime.parse(dates[1].to_rfc3339)
         )
-      end)
+
+        # See https://github.com/glacials/splits-io/pull/474/files#r241242051
+        dates.compact.each(&:dispose)
+
+        container << history
+        next if container.size < MAX_CONTAINER_SIZE
+
+        write_run_histories(container)
+        container = []
+      end
+
+      return if container.blank?
+
+      write_run_histories(container)
     end
 
-    def write_segments(parsed_segments)
-      segs = []
+    def write_run_histories(histories)
+      result = RunHistory.import(histories)
+      return result unless result.failed_instances.any?
+
+      raise UnparsableRun
+    end
 
-      parsed_segments.each do |parsed_segment|
-        segs << Segment.new(
+    def process_segments(run)
+      container = {}
+      previous_segment = OpenStruct.new(realtime_end_ms: 0, gametime_end_ms: 0)
+
+      (0...run.len).each do |i|
+        lsc_segment = run.segment(i)
+        segment = Segment.new(
           run_id:         id,
-          segment_number: parsed_segment[:segment_number],
-          name:           parsed_segment[:name],
-
-          realtime_start_ms:             parsed_segment[:realtime_start_ms], # starts can be 0 (for the first segment)
-          realtime_end_ms:               zero_to_nil(parsed_segment[:realtime_end_ms]),
-          realtime_duration_ms:          zero_to_nil(parsed_segment[:realtime_duration_ms]),
-          realtime_shortest_duration_ms: zero_to_nil(parsed_segment[:realtime_best_ms]),
-
-          realtime_skipped: parsed_segment[:realtime_skipped],
-          realtime_reduced: false,
-          realtime_gold:    parsed_segment[:realtime_gold],
-
-          gametime_start_ms:             parsed_segment[:gametime_start_ms], # starts can be 0 (for the first segment)
-          gametime_end_ms:               zero_to_nil(parsed_segment[:gametime_end_ms]),
-          gametime_duration_ms:          zero_to_nil(parsed_segment[:gametime_duration_ms]),
-          gametime_shortest_duration_ms: zero_to_nil(parsed_segment[:gametime_best_ms]),
-
-          gametime_skipped: parsed_segment[:gametime_skipped],
-          gametime_reduced: false,
-          gametime_gold:    parsed_segment[:gametime_gold]
+          segment_number: i,
+          name:           lsc_segment.name || '',
```

Comment:
This should fix segments with no name being rejected from the DB.

Reply:
👍 we might want to default this to something readable if we're not going to allow nulls in the db, like `segment_number` or `"???"` or something. Allowing blanks seems prone to UI bugs, e.g. a coaching situation where we say "You should improve your ."

I believe what we do for `run.user` is allow nulls but have `run.user.to_s` render as `"???"` so the UI still functions, but I wouldn't blame you if a migration for that felt out of scope for this PR.

Reply:
Do you think we should allow for nulls in the db for the name instead?  I don't really have a preference one way or the other, was just trying to keep things consistent with current behavior.

Reply:
I do think allowing nulls would be the cleanest, yeah. `null` represents not present, and that's the best description of what the user gave us for this field. Then override `Segment#name` to be like
```ruby
class Segment
  def name
    read_attribute(:name) || '???' # or '', or segment_number + 1, or 'Unnamed segment'
  end
end
```
but again, feel free to punt and hardcode something `'???'`-esque here in the parsing code if you feel that's out of scope for this PR.